specification.md

```diff
@@ -614,12 +620,14 @@ List<String>? getArguments()
 ```
 
 Sets/gets the argument list to be passed to the executable. Unlike with
-`execve()`, the first element of the list will correspond to `argv[1]`
-when accessed by the invoked executable. If no previous call to
-`setArguments` was made, `getArguments` will return `null`. The setter
-does not create a copy of the list. Therefore, it is possible to add
-arguments to the list by invoking `setArguments()` with a mutable list,
-then invoking `getArguments().add()`.
+`execve()`, the first element of the list will correspond to `argv[1]` when
+accessed by the invoked executable. If no previous call to `setArguments` was
+made, `getArguments` will return `null`. The setter does not create a copy of
+the list. Therefore, it is possible to add arguments to the list by invoking
+`setArguments()` with a mutable list, then invoking `getArguments().add()`. The
+strings in the argument list are subject to environment variable expansion, as
+described in
+[`JobSpecification.setEnvironment`](#jobspecification-setenvironment).
```

Comment:
Or maybe I misunderstand what `mutable` here means?

Reply:
A user could intentionally sabotage the process by modifying a spec (including but not limited to the argument list) while the job is being submitted, triggering a race condition. Unless we make the specs completely immutable, we won't get rid of this "problem". But I think many if not most APIs will be equally vulnerable.

In some cases, if the specs are populated from a thread different from the one in which submit() gets called, even if sequenced properly, some synchronization may be necessary to ensure that both threads have a consistent view of spec objects. This, I'd argue, is something a user has to do anyway when consuming, in a thread, objects generated in a different thread, so it won't be something specific to JPSI.

Reply:
I was worried about simple cases like this:
```py
args = ['foo', 'bar', -1]
for i in range(10):
    ...
    args[-1] = i
    jobspec.arguments = args
    ...
```
Would we expect to obtain 10 different or 10 identical jobs?  

You are probably right, it's likely not much of an issue in practice, such constructs are not really common.  So I don't mind leaving it as it is.  But it would also not hurt to remove the `mutable`, to allow implementations to make a deep copy.  It's not like the mutability is the goal, right?

Reply:
I think it's a pretty common scenario for which programmers need to be prepared.

I do think there are some trade-offs. We could instead:
- add more methods to allow manipulating arg lists
  - I don't like this on the grounds that we're mostly replicating `List` methods when most languages have perfectly reasonable list classes.
- make internal copies
  - this is probably a better solution, but it may make it harder to easily add arguments -- although that's quite debatable: one can call all the relevant list modification methods before doing `spec.setArguments(...)`

So maybe we should make internal copies.

Reply:
Yeah, trade-offs, as always...
I agree on not adding methods on list manipulation etc.  Brrr...  Requiring copies sounds clean, but now I can't make up my mind an am worried about the implied overhead in level 2, where we presumably want to target exascale (== many jobs)...

Reply:
Either the user makes the copies or the API does, but there will be copies.

Reply:
Pondered over this a bit more, and it's likely ok as is.  We can rediscuss if any issues emerge in practical use.